compiler/src/iree/compiler/Codegen/Utils/GPUUtils.cpp

```diff
@@ -280,7 +280,7 @@ void insertBarriersAroundSharedMemoryCopy(func::FuncOp funcOp) {
       Operation *prevOp = copyOp->getPrevNode();
       if (!prevOp || !hasMarker(prevOp, getCopyToWorkgroupMemoryMarker())) {
         builder.setInsertionPoint(copyOp);
-        builder.create<gpu::BarrierOp>(copyOp->getLoc());
+        // builder.create<gpu::BarrierOp>(copyOp->getLoc());
```

Comment:
what I meant is remove the change. Do you need this? I assumed this is a hack you forgot to remove?

Reply:
This will add a `bar.sync 0;` after every `cp.async` stage in the prologue. We don't need it for functionality, and it may add a performance overhead for matmuls with shorter mainloop a.k.a. smaller problem shape K. For performance, we really don't want to see `bar.sync 0;` in prologue for other than the last `cp.async` stage in the prologue. I have not quantified the performance penalty of keeping `bar.sync 0` `AroundSharedMemoryCopy` for every `cp.async` stage.

Reply:
right but this is not correct unless we multi-buffer and pipeline in a certain way so we cannot just remove it here. That would cause potential functionality problems.

Reply:
If we keep this change, i.e., have `gpu::BarrierOp` Around every SharedMemoryCopy. 

```mlir
nvgpu.gpu_barrier;               // barrier at the start of Smem copy
cp.async; // x4 for Tile A
cp.async; // x4 for Tile B
nvgpu.gpu_barrier;             // barrier at the end of Smem copy
```

Then the pipeline and schedule for Ampere Tensor Cores `getNvidiaAmpereTensorCorePipeline` should ignore the one at the start of Smem copy. It is not needed for multi-staged pipeline. I think it is there to support single-stage pipelines, where the same stage is read and written?

Reply:
>  I think it is there to support single-stage pipelines, where the same stage is read and written?

Yes. We can remove the barrier when we pipeline but removing it here is not correct.

Reply:
Removed the change. Thus, now we have two `gpu.gpu_barrier;`s when the IR hits scheduler/pipeliner, but we erase the `gpu. gpu_barrier ` that appears before the start of Shared Memory copy, as it is not required for the multi-staged Ampere-stype pipelines using `cp.async`